Use swagger-ui-dist

clarakosi opened the pull request:
Updated swaggerUI.js to use swagger-ui-dist. New feature:
request duration.

Bug: T218218

_______________________

~models: is turned on by default but can be turned off~ Turned off
request duration: I turned it on but can also remove

Pchelolo commented on lib/swaggerUI.js:
```diff
@@ -23,18 +23,25 @@ function staticServe(hyper, req) {
-                .replace(/ url: url,/, 'url: "?spec",');
+                .replace(/dom_id: '#swagger-ui'/, 'dom_id: "#swagger-ui", ' +
+                    'docExpansion: "list", validatorUrl: null, displayRequestDuration: true')
+                .replace(/"https:\/\/petstore.swagger.io\/v2\/swagger.json"/, `${req.headers.host}${req.uri}`);
```
You can still just use `"?spec" here to have a relative reference.

clarakosi replied:
Done

Pchelolo commented:
Couple of things for you to find out:
- Is it possible to make all the sections collapsed by default?
- Is it possible to hide the models? they're kinda useless.
- The descriptions of the responses preserve line breaks from the original YAML - is it possible to make it remove line breaks? if not, we will have to remove the line breaks in the spec.
- Is it possible to get rid of the authorize part?
- Is there maybe a way to have some of the input fields be text area instead of the text field? that would be nice for transform POST
- Will continue playing with it and looking for things to improve

clarakosi commented:
I'll look into the other items you mentioned

Pchelolo commented:
For minified .js and .css we also need to fix the sourceMap path, thus do something like body = body.toString().replace(/sourceMappingURL=/, 'sourceMappingURL=/?path=');

Pchelolo commented:
Seems like the new Swagger-UI also uses `data` url for some images. Let's allow it in the content-security-policy. Thus, `img-src 'self'` neds to become `img-src 'self' data:`

clarakosi commented:
ok

Pchelolo commented:
This also seems to be fixing the second item on https://phabricator.wikimedia.org/T125412

clarakosi commented:
> * Is it possible to make all the sections collapsed by default?

Yes & done.

> * Is it possible to hide the models? they're kinda useless.

Yes & done.

> * The descriptions of the responses preserve line breaks from the original YAML - is it possible to make it remove line breaks? if not, we will have to remove the line breaks in the spec.

Line breaks seems to be built in. Allegedly, we can turn it off (although I haven’t been able to recreate yet) an alternative would be like you said change the spec.
Discussed here:  https://github.com/swagger-api/swagger-ui/issues/4653

> * Is it possible to get rid of the authorize part?

Removed it via css. Done.

> * Is there maybe a way to have some of the input fields be text area instead of the text field? that would be nice for transform POST

From what I can tell right now if it’s in body it automatically creates a textarea. Will research more.

> * Is it possible to completely get rid of the 'explore' part of it? WE don't really want people to explore some other wiki docs from another wiki doc page. Makes no sense.

You want to remove just the button or the button + input field?

> * The general description looks kinda ugly. Are there any controls to change the font size etc or we need to write our own css?

Yeah I agree. You can only change it via CSS. I made some small alterations.

> For minified .js and .css we also need to fix the sourceMap path, thus do something like body = body.toString().replace(/sourceMappingURL=/, 'sourceMappingURL=/?path=');

I just looked and I’m not sure where exactly this is happening.

> Seems like the new Swagger-UI also uses data url for some images. Let's allow it in the content-security-policy. Thus, img-src 'self' neds to become img-src 'self' data:

Hmm this is the Swagger Logo in the nav bar so I'm not sure if we want it. Trying to mirror the original swagger ui, I removed it via CSS and replaced it with `cfg.ui_name`.

Pchelolo commented:
> Line breaks seems to be built in. Allegedly, we can turn it off (although I haven’t been able to recreate yet) an alternative would be like you said change the spec.
Discussed here: swagger-api/swagger-ui#4653

Ok, thank you for the investigation. Let's try to submit a pull request upstream removing that 'breaks: true' and see what happens? If we don't get any response in a timely manner, we can convert `|` to `>-` in our specs, but that's a less good solution..

 >  From what I can tell right now if it’s in body it automatically creates a textarea. Will research more.

It creates a text field, but we would like to be able to indicate that the input should be a text area. Not a big problem, but we could try opening a ticket upstream.

> You want to remove just the button or the button + input field?

The whole thing.

So, if you look into .js and .css files from swagger-ui-dist, in the bottom they have a comment with `sourceMappingURL`. This is pointing the browser to where the sourceMap file is located. The idea of the sourceMap is that you JS/CSS is served to the browser minified, but if the user wants to debug JS in the browser, it also loads a sourceMap file that has info on how to 'unminify' the sources. The location of this file is in the annotation in js/css files. Since we serve the contents of the swagger-ui-dist a bit differently, we need to apply the regex replacement I posted to .js and .css file contents.
See https://www.html5rocks.com/en/tutorials/developertools/sourcemaps/ for more details.

> Hmm this is the Swagger Logo in the nav bar so I'm not sure if we want it. Trying to mirror the original swagger ui, I removed it via CSS and replaced it with cfg.ui_name

It's still complainig about it, plus it wants to load some more pics. Let's allow it.

clarakosi commented:
> Ok, thank you for the investigation. Let's try to submit a pull request upstream removing that 'breaks: true' and see what happens? If we don't get any response in a timely manner, we can convert | to >-in our specs, but that's a less good solution..

Ok made a PR. 

> > You want to remove just the button or the button + input field?
> 
> The whole thing.

Done. 

> > Hmm this is the Swagger Logo in the nav bar so I'm not sure if we want it. Trying to mirror the original swagger ui, I removed it via CSS and replaced it with cfg.ui_name
> 
> It's still complainig about it, plus it wants to load some more pics. Let's allow it.

Done.

> The idea of the sourceMap is that you JS/CSS is served to the browser minified, but if the user wants to debug JS in the browser, it also loads a sourceMap file that has info on how to 'unminify' the sources.

For the html page we’ve been changing a copy but with this I would have to directly change the .js + .css files after install. This may be better suited for a PR upstream.

Pchelolo commented:
> Ok made a PR.

Could you reference the PR here?

> For the html page we’ve been changing a copy but with this I would have to directly change the .js + .css files after install.

Not really - we read the file in memory and changing things in memory too. Like in https://github.com/wikimedia/hyperswitch/blob/master/lib/swaggerUI.js#L53 for example. 

> This may be better suited for a PR upstream.
Unfortunately not. The sourceMappingURL is appended to the files during the webpack build step, thus we can't really make it configurable in runtime. The only option to change it upstream is to for swagger-ui-dist and rebuild the assets ourselves, which is much worse than running one regex in runtime.

clarakosi commented:
> > Ok made a PR.
> 
> Could you reference the PR here?
> 

https://github.com/swagger-api/swagger-ui/pull/5274
> > For the html page we’ve been changing a copy but with this I would have to directly change the .js + .css files after install.
> 
> Not really - we read the file in memory and changing things in memory too. Like in https://github.com/wikimedia/hyperswitch/blob/master/lib/swaggerUI.js#L53 for example.
> 

Ok. Will look into this.

clarakosi commented:
> > > For the html page we’ve been changing a copy but with this I would have to directly change the .js + .css files after install.
> > 
> > 
> > Not really - we read the file in memory and changing things in memory too. Like in https://github.com/wikimedia/hyperswitch/blob/master/lib/swaggerUI.js#L53 for example.
> 
> Ok. Will look into this.

Done

Pchelolo commented on lib/swaggerUI.js:
```diff
@@ -23,25 +23,42 @@ function staticServe(hyper, req) {
+                .scheme-container {
+                    display: none
+                }
+            `;
```
Could you please add a comment to each of the CSS blocks explaining what exactly it is doing? Will be beneficial later for maintenance.

clarakosi replied:
Done

Pchelolo commented on lib/swaggerUI.js:
```diff
@@ -23,25 +23,42 @@ function staticServe(hyper, req) {
-                .replace(/ url: url,/, 'url: "?spec",');
+                .replace(/dom_id: '#swagger-ui'/, 'dom_id: "#swagger-ui", ' +
+                    'docExpansion: "none", defaultModelsExpandDepth: -1, validatorUrl: null, displayRequestDuration: true')
+                .replace(/"https:\/\/petstore.swagger.io\/v2\/swagger.json"/, '?spec');
```
You forgot `"` in the replacement.

clarakosi replied:
Done

Pchelolo commented:
A couple of comments inlined and LGTM

Pchelolo commented:
Beautiful

Pchelolo commented:
Puclished as 0.11.0
